### PR TITLE
feat: update to go 1.27

### PR DESCRIPTION
**Pull request by rodrodros**
Updates the toolchain to Go 1.27 (go.mod, Dockerfile, docs, `starknet-p2p-specs/go.mod`, CI workflows) and consolidates the `felt.Slice` JSON v2 fast path into `core/felt/slice.go`, removing the `GOEXPERIMENT=jsonv2` guard job now that `encoding/json/v2` has graduated.

## Module-wide `encoding/json` v1 → v2 implementation switch

Per Go's GODEBUG compatibility rule, bumping the module's `go` directive to 1.27 switches the `encoding/json` **v1** API onto the v2 implementation for the **entire module** — every JSON decode in Juno (RPC handlers, feeder/gateway client, config), not only the `felt.Slice` fast path.

- `jsonrpc/pretty_error.go` is the most exposed spot: it inspects `*json.SyntaxError.Offset` and matches v1 parser error strings. Its tests pass under Go 1.27 and act as the regression pin for those RPC-facing error messages.
- Escape hatch if ever needed: `godebug jsonv1=1` in go.mod would decouple the toolchain bump from the JSON semantics switch. It is intentionally **not** used — the v2 fast path is the point of this change.

## User-visible error message change

`felt.Slice` decode errors for wrong-kind input now read:

```
cannot unmarshal slice: unexpected json kind: <kind>
```

replacing the previous `felt: cannot unmarshal ... into Slice` wording.

### CODE REVIEW

File: core/felt/slice_test.go [hunk 1]
```diff
@@ -337,6 +342,160 @@ func requireSliceDecodeJSONEquivalent(t *testing.T, data []byte) {
 	}
 }
 
+// requireSliceDecodeJSONv2Equivalent decodes data through the pure json/v2 API
+// with both the fast and generic paths and asserts they agree on acceptance,
+// on the decoded felts, and on whether the failure was a truncation.
+func requireSliceDecodeJSONv2Equivalent(t *testing.T, data []byte) {
```
Comment (RafaelGranza) [inline core/felt/slice_test.go:348]: I'm not sure why testing against the pure `json/v2` API.

No one calls the `json/v2` directly, only as a `json/v1` engine.

After moving to go 1.27, the current implementation of `requireSliceDecodeJSONEquivalent` already asserts the fast and generic should agree under the `json/v1`  with `json/v2` engine.

In that same direction, no  `requireSliceDecodeJSONv2Equivalent` call is necessary.

File: core/felt/slice_test.go [hunk 2]
```diff
@@ -337,6 +342,160 @@ func requireSliceDecodeJSONEquivalent(t *testing.T, data []byte) {
...
+			generic[idx],
+			fast[idx],
+		)
+	}
+}
+
+// Under the pure json/v2 API the nil-slice representation is governed by the
+// FormatNilSliceAsNull option ([] by default, null under v1-compat options),
+// and the fast path must stay byte-for-byte equivalent to the generic encoder.
+func TestSliceJSONv2Marshal(t *testing.T) {
```
Comment (RafaelGranza) [inline core/felt/slice_test.go:409]: Also no need to test pure `json/v2`.
`FormatNilSliceAsNull` is the default for `json/v1` with the `json/v2` engine.

I would delete this test.

File: core/felt/slice_test.go (hunk 1 above)
Comment (RafaelGranza) [reply core/felt/slice_test.go:348]: I would delete this helper.

File: core/felt/slice_test.go [hunk 3]
```diff
@@ -337,6 +342,160 @@ func requireSliceDecodeJSONEquivalent(t *testing.T, data []byte) {
...
+	asArray, err := jsonv2.Marshal(nilSlice)
+	require.NoError(t, err)
+	require.Equal(t, `[]`, string(asArray))
+
+	asNull, err := jsonv2.Marshal(nilSlice, jsonv2.FormatNilSliceAsNull(true))
+	require.NoError(t, err)
+	require.Equal(t, `null`, string(asNull))
+}
+
+func TestSliceJSONv2DecodeErrors(t *testing.T) {
```
Comment (RafaelGranza) [inline core/felt/slice_test.go:447]: I would also delete this one, based on the same criteria.

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module github.com/NethermindEth/juno
 
-go 1.26.0
+go 1.27.0
```
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply go.mod:3]: We understand enabling it touches the whole juno node.
That's why we are still using `json/v1`.

This way we only use `json/v2` as an engine for `json/v1`, and Go 1.27 guarantees that compatibility.

Related to `jsonrpc/pretty_error.go`, the change in behavior was a bug when enabling `json/v2` with Go 1.26. [It was fixed for Go 1.27!](https://github.com/golang/go/issues/79659)

File: core/felt/slice_test.go [hunk 5]
```diff
@@ -337,6 +342,160 @@ func requireSliceDecodeJSONEquivalent(t *testing.T, data []byte) {
...
+				requireSliceDecodeJSONv2Equivalent(t, []byte(input))
+			})
+		}
+	})
+}
+
+// The element-kind error can only come from UnmarshalJSONFrom, so this proves
+// at runtime that the fast path is wired in under both JSON APIs, complementing
+// the compile-time interface assertions in slice.go.
+func TestSliceJSONFastPathWiredIn(t *testing.T) {
```
Comment (RafaelGranza) [inline core/felt/slice_test.go:485]: I believe it is fair to have a test checking whether we are actually using the fast path. 
The main problem is that they are built to behave the same, so the only way to achieve this is through different errors in a distinguish corner case. Such as:

```go
func TestSliceJSONFastPathWiredIn(t *testing.T) {
	var out felt.Slice[felt.Felt]
	require.ErrorContains(t, json.Unmarshal([]byte(`[1]`), &out), "cannot unmarshal slice:")
}
```

On the other hand, the fast path only exists because it is supposed to be FASTER than the generic. So the way to assert this behavior is a benchmark, which is already implemented. See `BenchmarkSliceVsFeltArrayJSON` on `core/felt/slice_bench_test.go`

Comment (RafaelGranza) [reply core/felt/slice_test.go:485]: My suggestion is one of these:
- use a corner case error to assert
- delete it and go with the benchmark.

File: .github/workflows/juno-lint.yml [hunk 6]
```diff
@@ -27,6 +27,8 @@ jobs:
       # General Linter code wide
       - name: golangci-lint
         uses: golangci/golangci-lint-action@ba0d7d2ec06a0ea1cb5fa41b2e4a3ab91d21278a # v9.3.0
+        with:
+          version: latest
```
[the remark this answers is not in the document]
Comment (rodrodros) [reply .github/workflows/juno-lint.yml:31]: it is intentional

File: core/felt/slice.go [hunk 7]
```diff
@@ -128,3 +141,98 @@ func decodeCBORArrayHeader(data []byte) (size, offset int, ok bool) {
...
+		token, err := dec.ReadToken()
+		if err != nil {
+			return err
+		}
+		if token.Kind() == 'n' { // null -> nil slice
+			*s = nil
+			return nil
+		}
+		return fmt.Errorf("cannot unmarshal slice: unexpected json kind: %s", token.Kind())
+	}
```
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply core/felt/slice.go:199]: My suggestion here is removing the word "consumed" from the comment:
```go
// and a wrong-kind token is reported as one.
```

About `TestSliceJSONv2DecodeErrors`, it is not needed.

When moving to Go 1.27, the fast vs generic outputs are compared by both `requireSliceDecodeJSONEquivalent` using the new `json/v2` engine and the fuzz tests.

File: core/felt/slice.go [hunk 8]
```diff
@@ -128,3 +141,98 @@ func decodeCBORArrayHeader(data []byte) (size, offset int, ok bool) {
...
+	newSlice = slices.Grow(newSlice, bytes.Count(val, []byte{'"'})/2)
+
+	for index := 1; index < len(val)-1; {
+		switch val[index] {
+		case ' ', '\t', '\n', '\r', ',':
+			index++
+			continue
+		case '"':
+		default:
+			return fmt.Errorf("cannot unmarshal slice: unexpected json kind: %s", val[index:].Kind())
```
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply core/felt/slice.go:223]: IMO, i would keep the `felt:` prefix, only to be compatible with the errors on `core/felt/felt.go`.

But it is up to you.

**Review by RafaelGranza — COMMENTED** [review]
I believe it is correct, but it needs a scope alignment. (Specially for the reviewer bot)

We will not be using the `json/v2` directly, only `json/v1` with the `json/v2` engine.
We get the speed and keep compatibility, which Go 1.27 guarantees.

About `felt.Slice`, it must behave the same as `[]felt.Felt`.
Just from moving to Go 1.27, we will automatically start using `json/v2` engine with the new json marshaler and unmarshaler, so the old tests and benchmarks already assert the outcomes of the `generic` vs `fast path` are the same through the new methods.

To sum up, I don't think we need `json/v2` specific tests here, or any new tests at all. What we have already covers it.

File: core/felt/slice.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (rodrodros) [reply core/felt/slice.go:223]: The new unified, unprefixed error text is intentional. `TestSliceJSONFastPathWiredIn` remains sound even with both sites emitting the same message: for input `[1]` the top-level branch is unreachable (peek kind is `[`) and the generic reflection decoder never produces this text, so the assertion still proves `UnmarshalJSONFrom` runs under both JSON APIs. Test assertions have been updated to match the new messages.

File: core/felt/slice_test.go (hunk 2 above)
Comment (rodrodros) [reply core/felt/slice_test.go:409]: But we are using the Json v2 engine, with the default json v1 engine. I think this one is justified

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply go.mod:3]: I think the best approach is to wait for the next `codeql-cli` release, which should land in the next few days.

File: core/felt/slice.go [hunk 9]
```diff
@@ -128,3 +141,100 @@ func decodeCBORArrayHeader(data []byte) (size, offset int, ok bool) {
...
+		return valErr
+	}
+
+	newSlice := (*s)[:0]
+	if newSlice == nil {
+		newSlice = Slice[F]{}
+	}
+	// Each element contributes two quotes; escaped quotes never survive setHex,
+	// so a wrong hint only ever costs capacity, never correctness.
+	newSlice = slices.Grow(newSlice, bytes.Count(val, []byte{'"'})/2)
```
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply core/felt/slice.go]: makes sense, but I'd go with a const value:
```go
	minJSONFeltLen = len(`"0x0",`)
```

And:

```suggestion
        maxPossibleFelts := len(val) / minJSONFeltLen
        expectedNumberOfFelts := bytes.Count(val, []byte{'"'})/2
	newSlice = slices.Grow(newSlice, min(expectedNumberOfFelts, maxPossibleFelts))
```

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (RafaelGranza) [reply go.mod:3]: Main is already using the new version

File: core/felt/slice_test.go (hunk 2 above)
Comment (RafaelGranza) [reply core/felt/slice_test.go:409]: Now I understood. it is supposed to cover:

```go
if s == nil {
		if asNull, _ := json.GetOption(enc.Options(), json.FormatNilSliceAsNull); asNull {
			return enc.WriteToken(jsontext.Null)
		}
	}
```

A guard that only exists to cover a possible case when someone calls `marshal` through pure `json/v2`.

I would had deleted the guard, but since it is a `json/v2` interface, it is fair to treat it.

**Review by RafaelGranza — COMMENTED** [review]
In my last review, I considered the current tests fair.

Even thought we are not using pure `json/v2` in our codebase, since we are using new `json/v2` interfaces (`MarshalJSONTo` and `UnmarshalJSONFrom`), it is fair to have tests to assert they work on pure `json/v2`.

The only change I made was a rebase to include the new version of `codeql-actuon`and a limite to pre-allocated memory in `UnmarshalJSONFrom`.

**Review by RafaelGranza — APPROVED** [review]
